src/lib/integration_test_local_engine/mina_automation.ml

```diff
@@ -0,0 +1,462 @@
+            ; add_genesis_winner= None
+            ; num_accounts= None
+            ; balances= []
+            ; hash= None
+            ; name= None }
+      ; epoch_data= None }
+    in
+    let genesis_constants =
+      Or_error.ok_exn
+        (Genesis_ledger_helper.make_genesis_constants ~logger
+           ~default:Genesis_constants.compiled runtime_config)
+    in
+    let constants : Test_config.constants =
+      {constraints= constraint_constants; genesis= genesis_constants}
+    in
+    (* BLOCK PRODUCER CONFIG *)
+    let block_producer_config index keypair =
+      { name= "test-block-producer-" ^ Int.to_string (index + 1)
+      ; id= Int.to_string index
+      ; keypair
+      ; keypair_secret= keypair.secret_name
+      ; public_key= keypair.public_key_file
+      ; private_key= keypair.private_key_file
+      ; libp2p_secret= "" }
+    in
+    let block_producer_configs =
+      List.mapi block_producer_keypairs ~f:block_producer_config
+    in
+    let snark_coordinator_configs =
+      if num_snark_workers > 0 then
+        List.mapi
+          (List.init num_snark_workers ~f:(const 0))
+          ~f:(fun index _ ->
+            { name= "test-snark-worker-" ^ Int.to_string (index + 1)
+            ; id= Int.to_string index
+            ; snark_worker_fee
+            ; public_key= snark_worker_public_key } )
+        (* Add one snark coordinator for all workers *)
+        |> List.append
+             [ { name= "test-snark-coordinator"
+               ; id= "1"
+               ; snark_worker_fee
+               ; public_key= snark_worker_public_key } ]
+      else []
+    in
+    (* Combine configs for block producer configs and runtime config to be a docker bind volume *)
+    let docker_volume_configs =
+      List.map block_producer_configs ~f:(fun config ->
+          {name= "sk_" ^ config.name; data= config.private_key} )
+      @ [ { name= "runtime_config"
+          ; data=
+              Yojson.Safe.to_string (Runtime_config.to_yojson runtime_config)
+          } ]
+    in
+    let mina_archive_schema =
+      "https://raw.githubusercontent.com/MinaProtocol/mina/develop/src/app/archive/create_schema.sql"
+    in
+    { mina_automation_location= cli_inputs.coda_automation_location
+    ; debug_arg= debug
+    ; keypairs= block_producer_keypairs
+    ; constants
+    ; docker=
+        { version
+        ; stack_name
+        ; coda_image= images.coda
+        ; docker_volume_configs
+        ; runtime_config= Runtime_config.to_yojson runtime_config
+        ; block_producer_configs
+        ; snark_coordinator_configs
+        ; log_precomputed_blocks
+        ; archive_node_count= num_archive_nodes
+        ; mina_archive_schema } }
+
+  let to_docker network_config =
+    let open Docker_compose.Compose in
+    let open Node_constants in
+    let runtime_config = Service.Volume.create "runtime_config" in
+    let blocks_seed_map =
+      List.fold network_config.docker.block_producer_configs
```

Comment:
Instead of folding over empty maps to fill them up, you can also build an associative list and construct the map from that. Eg, `List.map ... ~f:(fun ... -> (k, v)) |> String.Map.of_alist`.